torch/testing/_internal/distributed/multi_threaded_pg.py

```diff
@@ -82,12 +82,14 @@ def __init__(self, op):
 
     @torch.no_grad()
     def work(self, data):
-        tensors = []
-        for src_rank in range(0, len(data)):
-            tensors.append(data[src_rank][0])
-        res = _reduce_ops[self.op](tensors)
-        for src_rank in range(len(data)):
-            data[src_rank][0].copy_(res)
+        for i in range(len(data[0])):
+            tensors = []
+            for src_rank in range(0, len(data)):
+                tensors.append(data[src_rank][i])
+            res = _reduce_ops[self.op](tensors)
```

Comment:
does this really work? i.e. `reduce_ops[self.op]` is `torch.sum`, this sum only apply to one tensor and can't append to all the tensors in the coalesing case right?

Reply:
This code is super confusing, it probably deserves some rewriting as it's like 3 levels of nested lists.

We're reducing all tensors in `data[*][i]`. The first dimension is rank.
We reduce input tensors separately.